### PR TITLE
number_formatter creation in the preview module now passes on the locale of Number objects to be more explict and try to fix the Windows bug in #453

### PR DESCRIPTION
**Pull request by palewire**
For #453

### CODE REVIEW

**Comment by onyxfish** [comment]
This looks good. Two minor tweaks:

* It needs a test, if one can be contrived. 
* We shouldn't override the `format` reserved keyword. `fmt` is fine.

Thanks for looking into this. It'll be good to have it solved before NICAR.

**Comment by palewire** [comment]
You are of course correct on both counts. I will do both of those things.

**Comment by palewire** [comment]
Moments ago we appllied this patch via copy/paste technology to the site-packages distributions of agate on a Windows machine that was experiencing the #453 bug. I can report it successfully solved the problem.

**Comment by onyxfish** [comment]
Excellent! Thanks for testing it!

**Comment by palewire** [comment]
I've appended two new commits to this pull request. The first fixes the variable namespace issue you identified with ``format``. The second one adds a Number with an non-English locale to the print table tests. 

The Number's locale feature and table printing tests are already pretty strong, so I'm not sure we need much more for this patch. The fact that existing tests continue to pass seems like a good indicator it's working. But it's possible I'm missing some obvious way it needs to be unittested.

**Comment by onyxfish** [comment]
I'd like to add a line to that test verifying that the output actually has a dutch-formatted number in it, otherwise I think the existing test is fine.

This also needs to be ported to `print_bars`, but I'm fine with @nbedi or I figuring that out when we merge it. :+1: on what we have here.

**Comment by palewire** [comment]
Got it. That better test is now appended to the pull request. 

How would ``print_bars`` porting work? I have to confess I've never used that feature.

**Comment by onyxfish** [comment]
It's pretty much the same logic. It also calls `format_decimal` and we just need to make sure the locale gets passed through.

**Comment by onyxfish** [comment]
I hate to backtrack on this now, but I just realized that this is actually not the right solution to this problem. Everywhere else in agate there is a strict separation between input parsing and output formatting. By passing the locale through from the type we've conflated those two things. Really we should be using whatever the user has specified as their system locale and then falling back to some safe default when that doesn't exist.

To put it another way, if I'm in the US, I should always see comma thousands separators in my output, even if I'm parsing a European file. And vice versa.

**Comment by onyxfish** [comment]
At a guess, without actually trying it, what we want is babel's [`default_locale`](http://babel.pocoo.org/en/latest/api/core.html#babel.core.default_locale). Something like:

```
format_decimal(v, format=fmt, locale=default_locale() or 'en_US')
```

**Comment by palewire** [comment]
Okie doke. If that's the aim, I think your obstacle is that the global "sniff" done by babel on [this line right here](https://github.com/python-babel/babel/blob/master/babel/numbers.py#L28) returns an empty string on Windows and then fails when that empty string is passed into the formatter [here](https://github.com/python-babel/babel/blob/master/babel/numbers.py#L226) as a keyword argument. 

If I understand things right, this patch works because it changes things so agate is always overriding that keyword variable with the ``Number``'s locale setting.

I don't have an opinion about how to fix it, other than it needs to be fixed.

This patch would definitely set a harder "default" for ``Number`` formats, and for the sake of helping Windows users it might default some Linux and OSX users to who have not explicitly cast their locales to English.

Though if you go the "system" route that you've specified, why is the locale on Number's set by default [as an hardcoded string](https://github.com/onyxfish/agate/blob/master/agate/data_types/number.py#L27) at all? Shouldn't that be sniffed off the system as well? 

Perhaps solving the root problem there with a cross-platform dynamic locale call and keeping this patch intact might make the turtles smile all the way down?

**Comment by palewire** [comment]
I see your second comment now. I hadn't considered that idea, but yeah that might work. I think it will remove the possibility of Number locale attributes from influencing outputs at all. Though maybe that's your aim!

**Comment by onyxfish** [comment]
Yep, that's the locale. Agate, shouldn't care at all about the formatting particulars of the data its processing. There are def some places where we could/should allow more overrides to support different formatting cases, but they should explicitly align with either input or output—not affect both.

As to whether or not `Number` should default to the system locale. I considered that and at the time I had concluded that most users were probably processing more "American numbers" than any other kind, no matter where they are in the world. In retrospect, I don't have much evidence for that. I'd be open to changing it if anyone finds it to be a problem.

**Comment by palewire** [comment]
Got it. I did not understand this separate between the ``Number`` and the formatter that you were aiming for when I drafted this pull request. Excuse the n00b. I can try to modify my patch to be more in-line with the proposal you made a couple comments up.

**Comment by onyxfish** [comment]
Thanks, Ben. Sorry it took me a few passes to recognize my own intentions. This stuff gets really thorny. If it wouldn't have made the API so much more complicated I would have had an entirely separate object for specifying input formatting and kept them out of the column types entirely.

**Comment by palewire** [comment]
I just pushed another approach at the patch that retrieves the default locale in the same manner as babel at the top of the file, defaulting back to English for the Windows fix. However, unless I misunderstand things, this patch alone will force the user to always print out in their system locale. So I also took the liberty of adding a new key word argument to ``print_table`` that allows the user to override that default.

**Comment by palewire** [comment]
Also, to be clear, I have not tested this patch on a Windows machine yet.

**Comment by onyxfish** [comment]
Ben, did you ever get a chance to test this fix on Windows?

**Comment by palewire** [comment]
No but I probably can when I'm in the office tomorrow.

**Comment by onyxfish** [comment]
That would be great! I'm hoping to push a release out this weekend.

**Comment by palewire** [comment]
Working on this now. To confirm, I just installed everything from scratch on my wife's Windows 8 laptop and ran the example.py script from the HEAD of your upstream master branch. Here is the error.

agate-error

Trying patch next.

**Comment by palewire** [comment]
I then checked out the ``windows-locale-fix`` submitted with this pull request and repeated the example.py command. Everything worked fine. So I believe this patch does indeed work.

agate-error

**Comment by palewire** [comment]
I followed up, just for fun by merging upstream/master into the test branch so that it was up to date with all of your other changes. example.py still ran without any trouble. 

Here's a curious sidenote. When I first started on this testing run, with a virgin Windows laptop, I followed my old Windows habits of yore and installed Python and git and all the dependencies using the Cygwyn shell that impersonates a UNIX-based system. Once I had it all installed, there was no bug and everything worked great. 

I was only able to replicate the bug and verify that the patch works by installing Python and agate in the standard Windows ``cmd`` shell. 

That doesn't matter in the end, but it does underscore that testing in a "pure" Windows environment is the only way to be really sure if you've nailed it.

**Comment by onyxfish** [comment]
Thanks for testing, Ben. That's great. Your last comments actually line up with what Matt Waite had found for another bug. Cygwin seems to set some additional environment variables that the vanilla shell doesn't.

Thanks again for testing. Merging this in now.
